sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -873,24 +873,30 @@ object InferFiltersFromGenerate extends Rule[LogicalPlan] {
       if !e.deterministic || e.children.forall(_.foldable) => generate
 
     case generate @ Generate(g, _, false, _, _, _) if canInferFilters(g) =>
-      // Exclude child's constraints to guarantee idempotency
-      val inferredFilters = ExpressionSet(
-        Seq(
-          GreaterThan(Size(g.children.head), Literal(0)),
-          IsNotNull(g.children.head)
-        )
-      ) -- generate.child.constraints
-
-      if (inferredFilters.nonEmpty) {
-        generate.copy(child = Filter(inferredFilters.reduce(And), generate.child))
-      } else {
-        generate
+      g.children.head match {
+        case _: CreateNonEmptyNonNullCollection =>
```

Comment:
In general, optimizer rules should be orthogonal.

For this case, I think it's better to add a new optimizer rule that removes the `IsNotNull` and size check predicates above  `CreateArray` and `CreateMap`.